contracts/test/StakeTableInvariant.t.sol

```diff
@@ -0,0 +1,341 @@
+// SPDX-License-Identifier: Unlicensed
+
+/* solhint-disable contract-name-camelcase, func-name-mixedcase, one-contract-per-file */
+
+pragma solidity ^0.8.0;
+// Libraries
+
+import "forge-std/Test.sol";
+
+using stdStorage for StdStorage;
+
+import { SafeTransferLib } from "solmate/utils/SafeTransferLib.sol";
+import { BN254 } from "bn254/BN254.sol";
+import { EdOnBN254 } from "../src/libraries/EdOnBn254.sol";
+import { AbstractStakeTable } from "../src/interfaces/AbstractStakeTable.sol";
+import { LightClient } from "../src/LightClient.sol";
+import { LightClientTest } from "../test/mocks/LightClientTest.sol";
+import { StakeTableCommonTest } from "../test/StakeTable.t.sol";
+
+// Token contract
+import { ExampleToken } from "../src/ExampleToken.sol";
+
+// Target contract
+import { StakeTable as S } from "../src/StakeTable.sol";
+
+contract StakeTableHandlerTest is StakeTableCommonTest {
+    S public stakeTable;
+    address public tokenCreator;
+    ExampleToken public token;
+    mapping(uint256 index => BN254.G2Point vk) public vks;
+    BN254.G2Point[] public vksWithdraw;
+    LightClientTest public lightClient;
+    address[] public users;
+    mapping(bytes32 vkHash => uint256 userIndex) public userIndexFromVk;
+
+    // Variables for testing invariant relative to Register
+    uint64 public nextRegistrationEpochBefore;
+    uint64 public pendingRegistrationsBefore;
+    uint64 public stakeTableFirstAvailableRegistrationEpoch;
+    uint64 public stakeTableNumPendingRegistrations;
+    bool public registrationCalledAtLeastOnce;
+
+    // Variables for testing invariant relative to requestExit and withdrawFunds
+    uint64 public nextExitEpochBefore;
+    uint64 public pendingExitsBefore;
+    uint64 public stakeTableFirstAvailableExitEpoch;
+    uint64 public stakeTableNumPendingExits;
+    uint64 public currentEpoch;
+    bool public requestExitCalledAtLeastOnce;
+
+    mapping(bytes32 blsKeyHash => uint64 exitEpoch) public exitEpochForBlsVK;
+    BN254.G2Point[] public requestExitKeys;
```

Comment:
if `requestExitKeys` -> `uint256[] public requestExitKeyIndexes`, then we wouldn't need to build `userIndexFromVk` map, instead, we would use the fuzzy-random index to select a exit key _index_, then use `vk[idx]` to get the vk, and use `exitEpochForBlsVk(vkHash)` to get exit epoch.

Reply:
I don't follow as `userIndexFromVk` is not used in the context of exit requests.

Reply:
Ah right, I understand now. Suggested change done in https://github.com/EspressoSystems/espresso-sequencer/pull/1008.